File: lab13/main.cpp
#include <iostream>
#include <cmath>

using namespace std;

#define n 5
#define v 20

int I(int eta) {
    return static_cast<int>(std::log2(eta) / 3) + 1;
}

int K(int eta, int i) {
    double result = 1;
    for (int j = 1; j < i; ++j) {
        result += static_cast<double>(eta) / std::pow(8, j);
    }
    return static_cast<int>(result);
}

double eta2KFunc(int eta) {
    return static_cast<double>(eta) * std::log2(static_cast<double>(eta));
}

double NkFunc(double eta) {
    return 2 * eta * std::log2(eta);
}

double NFunc(int k, double Nk) {
    return static_cast<double>(k) * Nk;
}

double VFunc(int k, double Nk, double eta) {
    return static_cast<double>(k) * Nk * std::log2(2 * eta);
}

double PFunc(double N) {
    return 3 * N / 8;
}

double TFunc(double P) {
    return P / (static_cast<double>(n) * v);
}

double BFunc(double V) {
    return V / 3000;
}

double tauFunc(double T) {
    return T / 2;
}

double tnFunc(double tau, double B) {
    return tau / std::log(B);
}

int main()
{
    int eta[] = {8, 300, 400, 512};

    for (const auto& it : eta) {
        cout << "===============================================\n";
        int i = I(it);
        int k = K(it, i);
        double eta2k = eta2KFunc(it);
        double Nk = NkFunc(eta2k);
        double N = NFunc(k, Nk);
        double V = VFunc(k, Nk, eta2k);
        double P = PFunc(N);
        double T = TFunc(P);
        double B = BFunc(V);
        double tau = tauFunc(T);
        double tn = tnFunc(tau, B);
        
        cout << "eta = " << it << endl;
        cout << "i = " << i << endl;
        cout << "k = " << k << endl;
        cout << "eta2k = " << eta2k << endl;
        cout << "Nk = " << Nk << endl;
        cout << "N = " << N << endl;
        cout << "V = " << V << endl;
        cout << "P = " << P << endl;
        cout << "T = " << T << endl;
        cout << "B = " << B << endl;
        cout << "tau = " << tau << endl;
        cout << "tn = " << tn << endl;
    }

    return 0;
}

File: lab13/main.py
import math

def I(eta):
    return math.log2(eta) / 3 + 1

def K(eta, i):
    result = 1
    for j in range(1, i):
        result += eta / (8 ** j)
    return result

def eta2KFunc(eta):
    return eta * math.log2(eta)

def NkFunc(eta):
    return 2 * eta * math.log2(eta)

def NFunc(k, Nk):
    return k * Nk

def VFunc(k, Nk, eta):
    return k * Nk * math.log2(2 * eta)

def PFunc(N):
    return 3 * N / 8

def TFunc(P):
    return P / (5 * 20)

def BFunc(V):
    return V / 3000

def tauFunc(T):
    return T / 2

def tnFunc(tau, B):
    return tau / math.log(B)

def main():
    eta_values = [8, 300, 400, 512]
    
    for eta in eta_values:
        print("")
        i = I(eta)
        k = K(eta, int(i))
        eta2k = eta2KFunc(eta)
        Nk = NkFunc(eta2k)
        N = NFunc(k, Nk)
        V = VFunc(k, Nk, eta2k)
        P = PFunc(N)
        T = TFunc(P)
        B = BFunc(V)
        tau = tauFunc(T)
        tn = tnFunc(tau, B)
        
        print(f"eta = {eta}")
        print(f"i = {i}")
        print(f"k = {k}")
        print(f"Nk = {Nk}")
        print(f"N = {N}")
        print(f"V = {V}")
        print(f"P = {P}")
        print(f"T = {T}")
        print(f"B = {B}")
        print(f"tau = {tau}")
        print(f"tn = {tn}")

if __name__ == "__main__":
    main()
